src/aws-util/src/s3_uploader.rs

```diff
@@ -0,0 +1,554 @@
+///
+/// From <https://docs.aws.amazon.com/AmazonS3/latest/userguide/qfacts.html>
+const AWS_S3_MAX_PART_COUNT: i32 = 10_000;
+/// The minimum size of a part in a multipart upload.
+///
+/// This minimum doesn't apply to the last chunk, which can be any size.
+///
+/// From <https://docs.aws.amazon.com/AmazonS3/latest/userguide/qfacts.html>
+const AWS_S3_MIN_PART_SIZE: ByteSize = ByteSize::mib(5);
+/// The maximum size of a part in a multipart upload.
+///
+/// From <https://docs.aws.amazon.com/AmazonS3/latest/userguide/qfacts.html>
+const AWS_S3_MAX_PART_SIZE: ByteSize = ByteSize::gib(5);
+/// The maximum size of an object in s3.
+///
+/// From <https://docs.aws.amazon.com/AmazonS3/latest/userguide/qfacts.html>
+const AWS_S3_MAX_OBJECT_SIZE: ByteSize = ByteSize::tib(5);
+
+/// Information for an in-progress multi part upload after each `add_chunk` call.
+#[derive(Debug)]
+pub struct InProgressUpload {
+    /// The number of parts uploaded till now.
+    pub part_count: u32,
+    /// The number of bytes uploaded for all parts till now.
+    pub bytes_uploaded: u64,
+    /// The number of bytes remaining in the buffer which has not yet been uploaded.
+    pub bytes_buffered: u64,
+}
+
+/// Information about a completed multi part upload after `finish` is called.
+#[derive(Debug)]
+pub struct CompletedUpload {
+    /// The total number of parts in the multi part upload.
+    pub part_count: u32,
+    /// The total number of bytes uploaded in the multi part upload.
+    pub total_bytes_uploaded: u64,
+}
+
+/// Configuration object to configure the behaviour of the `S3MultiPartUploader`.
+/// Should be initialized by either calling `S3MultiPartUploaderConfig::try_new`
+/// or use `Default::default` to get pre-configured default values.
+#[derive(Debug)]
+pub struct S3MultiPartUploaderConfig {
+    part_size_limit: u64,
+    file_size_limit: u64,
+}
+
+impl S3MultiPartUploaderConfig {
+    /// Choosing a reasonable default for the maximum file size which
+    /// this uploader can upload. This can be overridden in the
+    /// `S3MultiPartUploaderConfig` config.
+    const DEFAULT_MAX_FILE_SIZE: ByteSize = ByteSize::gib(5);
+    /// Choosing a reasonable default for a part size. This can be overridden in the
+    /// `S3MultiPartUploaderConfig` config.
+    const DEFAULT_PART_SIZE_LIMIT: ByteSize = ByteSize::mib(10);
+
+    /// Returns an instance of a `S3MultiPartUploaderConfig`.
+    /// The given `part_size_limit` is how much data would be kept in the buffer of `S3MultiPartUploader`
+    /// before uploading a part.
+    /// As per S3 limits, the part size cannot be less than 5MiB and cannot exceed 5GiB.
+    /// The given `file_size_limit` is the maximum configured limit for the current object getting uploaded.
+    /// As per S3 limits, the object size cannot exceed 5TiB.
+    pub fn try_new(
+        part_size_limit: u64,
+        file_size_limit: u64,
+    ) -> Result<S3MultiPartUploaderConfig, anyhow::Error> {
```

Comment:
Minor Rust API design suggestion: `S3MultiPartUploaderConfig::try_new(1024, 2048)` is not very self describing as an API, because there's no natural order to indicate whether it's (part size, file size) or (file size, part size). I'd recommend making both fields public, and doing the validation of the fields inside of `S3MultiPartUploader::try_new`, rather than having a dedicated `S3MultiPartUploaderConfig::try_new` method.

Then the API the user interacts with switches from:

```
let config = S3MultiPartUploaderConfig::try_new(1024, 2048)?;
S3MultiPartUploader::try_new(sdk_config, bucket, path, config)?;
```

to:

```rust
S3MultiPartUploader::try_new(sdk_config, bucket, path, S3MultiPartUploaderConfig {
    part_size_limit: 1024,
    file_size_limit: 2048,
})?;
```

That reads quite nicely, and only having one fallible method to deal with is often easier to deal with!

Reply:
Ack. I was thinking in terms of individual constructors, but this makes sense.